doc/monitoring_containers/monitoring_containers.md

```diff
@@ -0,0 +1,276 @@
+  the resource usage of a docker container is beyond the pre-defined threshold.
+2. We leveraged the mechanism of event listener in supervisord to auto-restart a docker container
+  if one of its critical processes exited unexpectedly. 
+3. We also added a knob to make this auto-restart feature dynamically configurable.
+  Specifically users can run CLI to configure this feature residing in Config_DB as
+  enabled/disabled status.
+
+## 1.1 Requirements
+
+### 1.1.1 Functional Requirements
+1. The Monit must provide the ability to generate an alert when a critical process is not
+    running.
+2. The Monit must provide the ability to generate an alert when the resource usage of
+    a docker contaier is larger than the pre-defined threshold.
+3. The event listener in supervisord must receive the signal when a critical process in 
+    a docker container crashed or exited unexpectedly and then restart this docker 
+    container.
+4. CONFIG_DB can be configured to enable/disable this auto-restart feature for each docker
+    container.. 
+5. Users can access this auto-restart information via the CLI utility
+    1. Users can see current auto-restart status for docker containers.
+    2. Users can configure auto-restart status for a specific docker container.
+
+### 1.1.2 Configuration and Management Requirements
+Configuration of the auto-restart feature can be done via:
+1. init_cfg.json
+2. CLI
+
+### 1.1.3 Scalability Requirements
+`Place holder`
+
+## 1.2 Design
+
+### 1.2.1 Basic Approach
+Monitoring the running status of critical processes and resource usage of docker containers
+are heavily depended on the Monit system tool. Since Monit already provided the mechanism
+to check whether a process is running or not, it will be straightforward to integrate this into monitoring 
+the critical processes in SONiC. However, Monit only gives the method to monitor the resource
+usage per process level not container level. As such, monitoring the resource usage of a docker 
+container will be an interesting and challenging problem. In our design, we adopted the way
+that Monit will check the returned value of a script which reads the resource usage of docker 
+container, compares it with pre-defined threshold and then exited. 
+
+We employed the mechanism of event listener in supervisord to achieve auto-restarting of docker 
+container. Currently supervisord will monitor the running status of each process in SONiC
+docker containers. If one critical process exited unexpectedly, supervisord will catch such signal
+and send it to event listener. Then event listener will kill the process supervisord and
+the entire docker container will be shut down and restarted.
+
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+This feature is used to perform the following functions:
+1. Monit will write an alert message into syslog if one if critical process exited unexpectedly.
```

Comment:
This isn't how it works. Monit will write a message to syslog if a critical process has not been alive for 5 minutes.

Reply:
Reworded.